Public/Src/Tools/SandboxedProcessExecutor/OutputErrorObserver.cs

```diff
@@ -0,0 +1,126 @@
+            if (line.Length > 400)
+            {
+                line = line.Substring(0, 400);
+            }
+
+            // If a tool has a large amount of output that isn't a warning (eg., "dir /s %hugetree%")
+            // the regex matching below may be slow. It's faster to pre-scan for "warning"
+            // and bail out if neither are present.
+            if (m_isDefaultWarningRegex &&
+                line.IndexOf("warning", StringComparison.OrdinalIgnoreCase) == -1)
+            {
+                return false;
+            }
+
+            return m_warningRegex.IsMatch(line);
+        }
+
+        /// <summary>
+        /// Creates an instance of <see cref="OutputErrorObserver"/>
+        /// </summary>
+        public static OutputErrorObserver Create(SandboxedProcessInfo info)
+        {
+            Contract.Requires(info != null);
+
+            Regex warningRegex = info.StandardObserverDescriptor == null || info.StandardObserverDescriptor.WarningRegex == null
+                ? null
+                : CreateRegex(info.StandardObserverDescriptor.WarningRegex);
+            bool isDefaultWarningRegex = warningRegex == null ? false : IsDefaultWarningRegex(info.StandardObserverDescriptor.WarningRegex);
+            bool logOutputToConsole = info.StandardObserverDescriptor == null ? false : info.StandardObserverDescriptor.LogOutputToConsole;
+            bool logErrorToConsole = info.StandardObserverDescriptor == null ? false : info.StandardObserverDescriptor.LogErrorToConsole;
+
+            return new OutputErrorObserver(warningRegex, isDefaultWarningRegex, logOutputToConsole, logErrorToConsole);
+        }
+
+        private static bool IsDefaultWarningRegex(SandboxedProcessInfo.RegexDescriptor regexDescriptor) 
+            => string.Equals(regexDescriptor.Pattern, Warning.DefaultWarningPattern) && regexDescriptor.Options == RegexOptions.IgnoreCase;
+
+        private static Regex CreateRegex(SandboxedProcessInfo.RegexDescriptor regexDescriptor) 
+            => new Regex(
+                    regexDescriptor.Pattern,
+                    regexDescriptor.Options | RegexOptions.Compiled | RegexOptions.CultureInvariant);
+
+        /// <summary>
+        /// Observes output for warning.
+        /// </summary>
+        public void ObserveOutput(string outputLine)
+        {
+            if (IsWarning(outputLine))
+            {
+                Interlocked.Increment(ref m_warningCount);
+            }
+
+            if (m_logOutputToConsole)
+            {
+                Console.Out.WriteLine(outputLine);
+            }
+        }
+
+        /// <summary>
+        /// Observes error for warning.
+        /// </summary>
+        public void ObserveError(string errorLine)
+        {
+            if (IsWarning(errorLine))
+            {
+                Interlocked.Increment(ref m_warningCount);
+            }
+
+            if (m_logErrorToConsole)
+            {
+                Console.Error.WriteLine(errorLine);

```

Comment:
>e.Error.WriteLine(e [](start = 22, length = 19)

Shouldn't this be colored as well like in Executor?

Reply:
Yes. Danny also spotted that.

---
In reply to: [276826980](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276826980) [](ancestors = 276826980)

Reply:
Cool, didn't know I split in two :)

---
In reply to: [276874292](https://github.com/Microsoft/BuildXL/pull/159#discussion_r276874292) [](ancestors = 276874292,276826980)